### PR TITLE
Improve token parser

### PR DESCRIPTION
**Pull request by zurk**
- Add small subtokens not only to the next token but also to the previous:
before:
`'sourced.ml.algorithms.uast_ids_to_bag' -> ['sourc', 'algorithm', 'mlalgorithm', 'uast', 'ids', 'bag', 'tobag']`
after: 
`'sourced.ml.algorithms.uast_ids_to_bag' -> ["sourc", "sourcedml", "algorithm", "mlalgorithm", "uast", "ids", "idsto", "bag", "tobag"]`
- Add test cases
- refactor `prev_p[0]` to `ret.prev_p`

### CODE REVIEW

File: sourced/ml/tests/test_token_parser.py [hunk 1]
```diff
@@ -24,7 +32,8 @@ def test_split(self):
         self.assertEqual(list(self.tp.split("PrintAllExcept")), ["print", "all", "except"])
         self.assertEqual(
             list(self.tp.split("print really long line")),
-            ["print", "really", "long"])
+            # 'longli' is expected artifact due to edge effects
+            ["print", "really", "long", "longli"])
```
Comment (vmarkovtsev) [inline sourced/ml/tests/test_token_parser.py]: Please add more tests

**Comment by vmarkovtsev** [comment]
@zurk ping

**Comment by zurk** [comment]
What kind of tests do you mean?
Just more examples?

**Comment by vmarkovtsev** [comment]
Yes, all sorts of the craziest identifiers you can imagine.

File: sourced/ml/tests/test_token_parser.py [hunk 2]
```diff
@@ -9,19 +9,30 @@ def setUp(self):
...
-                         ["sourc", "sourcedml", "algorithm", "mlalgorithm",
-                          "uast", "ids", "idsto", "bag", "tobag"])
-        self.assertEqual(list(self.tp.process_token("a.b.c.d")), ["a", "b", "c", "d"])
 
+        tokens = [
+            ("sourced.ml.algorithms.uast_ids_to_bag",
+             ["sourc", "sourcedml", "algorithm", "mlalgorithm",
+              "uast", "ids", "idsto", "bag", "tobag"]),
+            # Bad example. Parser failed to parse it correctly
+            ("WORSTnameYOUcanIMAGINE", ['worst', 'name', 'nameyo', 'ucan', 'youcan', 'imagin']),
```
Comment (vmarkovtsev) [inline sourced/ml/tests/test_token_parser.py]: This is too bad ^, need to fix the parser on this example.

**Comment by vmarkovtsev** [comment]
@zurk Ping here, I want this change because it affects Waren's splitter

**Comment by zurk** [comment]
ok, Maybe, it is not a good time to fix this example? because also I could use imagine something like:
`WORSTNameYOUCanIMAGINE` and everything collides... 

My suggestion leave it as is for now.

**Comment by vmarkovtsev** [comment]
All right, please promise to finish this after 15th :smile:

**Comment by zurk** [comment]
I promise :)
Do You leave PR open for now?

**Comment by vmarkovtsev** [comment]
Yes.

**Comment by zurk** [comment]
So, @vmarkovtsev what should we do with the following two examples?
`
WORSTNameYOUCanIMAGINE
`
and (watch for caps)
`
WORSTnameYOUcanIMAGINE
`
should one of this be considered as not so mad and be fixed?

right now we have next splitting:
`WORSTNameYOUCanIMAGINE` to `['worst', 'name', 'ucan', 'imagin']`
`WORSTnameYOUcanIMAGINE` to `['worstn', 'ame', 'you', 'can', 'imagin']`

My idea that they are equal and fix for one broke another and vice versa.

Nerual splitter solves all this problem :)

**Comment by vmarkovtsev** [comment]
Now that we have our neural splitter, it is much easier indeed. Still, I expect

```
"WORSTnameYOUcanIMAGINE": ["worst", "name", "you", "can", "imagine"]
```

with a few alternative chunks at worst.

**Comment by zurk** [comment]
Now it should work

**Pull request merged**